### PR TITLE
add locale file canonicalization script

### PR DESCRIPTION
**Pull request by kevinkreiser**
@noblige we liked the idea of standardizing the formatting for the localization files. we did want to do a little more than just a strict alphabetical sort though so that we could maintain parts of manual organization for the sake of readability for contributor.

### CODE REVIEW

**Comment by noblige** [comment]
looks like this messes up unicode:
```
-        "1 quilòmetre",
+        "1 quil\u00f2metre", 
```

File: locales/canonicalize.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+def sort_dict(parent, d):
+  ordered = OrderedDict()
+  for k, v in sorted(d.iteritems(), key = lambda(k, v): sort_order(parent, k)):
+    if isinstance(v, dict):
+      ordered[k] = sort_dict(k, v)
+    else:
+      ordered[k] = v
+  return ordered
+
+#open the original file and sort it
```
Comment (noblige) [inline locales/canonicalize.py]:
```
 #open the original file and sort it
-with open(sys.argv[1], 'r') as f:
+with codecs.open(sys.argv[1], encoding='utf-8') as f:
   lang = sort_dict(None, json.load(f))
 
 #format it and write it back out
-with open(sys.argv[1], 'w') as f:
-  json.dump(lang, f, indent=2)
+with codecs.open(sys.argv[1], encoding='utf-8', mode='w') as f:
+   json.dump(lang, f, ensure_ascii=False, indent=2)
```

**Comment by kevinkreiser** [comment]
oh yeah i had tried that locally too out of worry but for some reason
forgot to put that change in this pr. i'll put it in thanks

2017-06-23 15:50 GMT-04:00 Aleksei Potov <notifications@github.com>:

**Comment by noblige** [comment]
@kevinkreiser another issue is that this sort is not stable, so this script produces different order for me, I think problem is here:
`  return order.index('SPOT_FOR_TAGS') if parent and not parent.isdigit() and parent not in order else k`

**Comment by noblige** [comment]
Sorry, I meant problem is that we have our JSON already sorted, and this script maintains previous order for tags not in the list (`order`).
I wonder if we can rework this to sort alphabetically keys that are not in the list, instead of maintaining their previous order?

**Comment by noblige** [comment]
Specific example, from en-US:
```
"phrases": {
        "0": "Bear <RELATIVE_DIRECTION>.", 
        "1": "Bear <RELATIVE_DIRECTION> onto <STREET_NAMES>.", 
        "2": "Bear <RELATIVE_DIRECTION> onto <BEGIN_STREET_NAMES>. Continue on <STREET_NAMES>.", 
        "3": "Bear <RELATIVE_DIRECTION> to stay on <STREET_NAMES>."
      }, 
      "relative_directions": [
        "left", 
        "right"
      ], 
      "empty_street_name_labels": [
        "the walkway", 
        "the cycleway", 
        "the mountain bike trail"
      ], 
      "example_phrases": {
        "0": [
          "Bear right."
        ], 
        "1": [
          "Bear left onto Arlen Road."
        ], 
        "2": [
          "Bear right onto Belair Road/US 1 Business. Continue on US 1 Business."
        ], 
        "3": [
          "Bear left to stay on US 15 South."
        ]
      }
```
`relative_directions` should be after `empty_street_name_labels`

**Comment by kevinkreiser** [comment]
@noblige  ok ill have a look!

**Comment by noblige** [comment]
proposed fix:
```
index 00724217..4a8edaed 100755
--- a/locales/canonicalize.py
+++ b/locales/canonicalize.py
@@ -22,7 +22,7 @@ def sort_order(parent, k):
 #and while doing so sort each subdict using the sorter above
 def sort_dict(parent, d):
   ordered = OrderedDict()
-  for k, v in sorted(d.iteritems(), key = lambda(k, v): sort_order(parent, k)):
+  for k, v in sorted(sorted(d.iteritems()), key = lambda(k, v): sort_order(parent, k)):
```

**Pull request merged**